src/transformers/adapters/peft_mixin.py

```diff
@@ -0,0 +1,300 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import inspect
+from typing import Optional
+
+from ..utils import (
+    check_peft_version,
+    find_adapter_config_file,
+    is_accelerate_available,
+    is_peft_available,
+    logging,
+    requires_backends,
+)
+
+
+if is_accelerate_available():
+    from accelerate import dispatch_model
+    from accelerate.utils import get_balanced_memory, infer_auto_device_map
+
+
+logger = logging.get_logger(__name__)
+
+
+class PeftAdapterMixin:
+    """
+    A class containing all functions for loading and using adapters weights that are supported in PEFT library.
+    Currently supported PEFT methods are all non-prefix tuning methods
+    """
+
+    _hf_peft_config_loaded = False
+
+    @check_peft_version(min_version="0.4.0")
```

Comment:
Instead of adding a decorator to each method, I would just use a different function than `requires_backends` in each function.

Reply:
Done in [ec51272](https://github.com/huggingface/transformers/pull/25077/commits/ec51272a07c66a5c7920616901d1c5a08826e123)
just for learning purpose, why adding a context manager is not preferred over a simple method? Maybe code readability purpose?